tutorials/source-modeling/plot_beamformer_lcmv.py

```diff
@@ -0,0 +1,207 @@
+# Introduction to beamformers
+# ---------------------------
+# A beamformer is a spatial filter that reconstructs source activity by
+# scanning through a grid of pre-defined source points and estimating activity
+# at each of those source points independently. A set of weights is
+# constructed for each defined source location which defines the contribution
+# of each sensor to this source.
+# Beamformers are often used for their focal reconstructions and their ability
+# to reconstruct deeper sources. They can also suppress external noise sources.
+# The beamforming method applied in this tutorial is the linearly constrained
+# minimum variance (LCMV) beamformer [1]_ which operates on time series.
+# Frequency-resolved data can be reconstructed with the dynamic imaging of
+# coherent sources (DICS) beamforming method [2]_.
+# As we will see in the following, the spatial filter is computed from two
+# ingredients: the forward model solution and the covariance matrix of the
+# data.
+
+###############################################################################
+# Data processing
+# ---------------
+# We will use the sample data set for this tutorial and reconstruct source
+# activity on the trials with left auditory stimulation.
+
+data_path = sample.data_path()
+subjects_dir = data_path + '/subjects'
+raw_fname = data_path + '/MEG/sample/sample_audvis_filt-0-40_raw.fif'
+
+# Read the raw data
+raw = mne.io.read_raw_fif(raw_fname, preload=True)
+raw.info['bads'] = ['MEG 2443']  # bad MEG channel
+
+# Set up the epoching
+event_id = 1  # those are the trials with left-ear auditory stimuli
+tmin, tmax = -0.2, 0.5
+events = mne.find_events(raw)
+
+# pick relevant channels
+raw.pick(['meg', 'eog'])  # pick channels of interest
+
+# Create epochs
+proj = False  # already applied
+epochs = mne.Epochs(raw, events, event_id, tmin, tmax,
+                    baseline=(None, 0), preload=True, proj=proj,
+                    reject=dict(grad=4000e-13, mag=4e-12, eog=150e-6))
+
+# Visualize averaged sensor space data
+evoked = epochs.average()
+evoked.plot_joint()
+
+###############################################################################
+# Computation of covariance matrices
+# ----------------------------------
+# Spatial filters are using the data covariance in the estimation of the filter
+# weights. The data covariance matrix will be `inverted`_ during the spatial
+# filter computation, so it is valuable to plot the covariance matrix and its
+# eigenvalues to gauge whether matrix inversion will be possible.
+# Also, because we want to combine different channel types (magnetometers and
+# gradiometers), we need to account for the different amplitude scales of these
+# channel types. To do this we will supply a noise covariance matrix to the
+# beamformer, which will be used for whitening.
+# The data covariance matrix should be estimated from a relevant time window
+# and incorporate enough samples for a stable estimate. Here, we use a time
```

Comment:
Well, MNE-Python actually will throw a warning (but I don't know at which threshold). There are rules of thumb out there (also published) but in the end it is the question whether the source reconstruction looks weird or not ... and whether the covariance matrix shows the typical pattern - but I am not sure how to explain this. Ideas?

Reply:
I looked through the `compute_covariance` code but didn't easily find a "not enough samples" warning, so it must be buried a few layers deep (perhaps in scipy or sklearn code that we call).  So I'd say, if there are published rules of thumb, I think it would be good to at least mention them.  Something like "...and incorporates enough samples for a stable estimate (XXX is a rule of thumb for the minimum number of samples, according to YYY)."  WDYT?

Reply:
So Brookes et al. 2008 say a useful rule of thumb "is that the number of samples _N_ must be greater than the total number of channels _M_". They, however, also discuss to do several beamformer estimates with several window lengths - and if the beamformer estimate changes, one operates "on the slope of [the] power curves", i.e., has not a stable covariance estimate.

Ref: Brookes et al., 2008, _Optimising experimental design for MEG beamformer imaging_, NeuroImage 39, 1788-1802.

Reply:
Maybe then we could say "...enough samples for a stable estimate. A good rule of thumb is to use at minimum as many time samples as there are channels in the data; see [3]_ for more detailed advice on covariance estimation for beamformers."

Reply:
Done!